src/swarm-lang/Swarm/Language/Syntax.hs

```diff
@@ -990,6 +998,76 @@ constInfo c = case c of
   lowShow :: Show a => a -> Text
   lowShow a = toLower (from (show a))
 
+------------------------------------------------------------
+-- SrcLoc
+------------------------------------------------------------
+
+data SrcLoc
+  = NoLoc
+  | -- | Half-open interval from start (inclusive) to end (exclusive)
+    SrcLoc Int Int
+  deriving (Eq, Ord, Show, Data, Generic, FromJSON, ToJSON)
+
+instance Semigroup SrcLoc where
+  NoLoc <> l = l
+  l <> NoLoc = l
+  SrcLoc s1 e1 <> SrcLoc s2 e2 = SrcLoc (min s1 s2) (max e1 e2)
+
+instance Monoid SrcLoc where
+  mempty = NoLoc
+
+-- | Check whether one SrcLoc starts before another one.
+srcLocBefore :: SrcLoc -> SrcLoc -> Bool
+srcLocBefore (SrcLoc a _) (SrcLoc b _) = a <= b
```

Comment:
Justify choice of `<=` vs `<`?

Reply:
I added a comment to clarify.  The fact is it doesn't really matter in practice, since we only ever use this to compare `SrcLoc`s between comments and code, which cannot possibly start at the same index.